### PR TITLE
Adding migration script to convert to utf8mb4

### LINKED ISSUE #989: Database seems to be latin1, need to convert to utf8mb4
Thank you for contributing to this project!

- Make sure to search the [issues](https://github.com/tl-its-umich-edu/my-learning-analytics/issues) for duplicates first!

## Expected behavior (A clear and concise description of what you expected to happen) : 
When looking at the database all of the tables and columns are latin1.

We should generate a conversion, add the database option for UTF8 and have a migration for this.

This could be run similar to this script.
https://stackoverflow.com/a/50935608/3708872

I'll work on this next week.

### PR DESCRIPTION
**Pull request by jonespm**
Fixes #989

### CODE REVIEW

**Comment by jonespm** [comment]
This seems to *work*

This was generated partially generated by the script at https://gist.github.com/jonespm/7e26665e4b489c4ebaef79f74100cb26

I took that, removed the database name (since it was going to operate on the current database) and broke it up into pieces. 

I tried to use https://pypi.org/project/django-mysql-utf8mb4/ but I couldn't get it to work right for me. 

This seems like something everyone will want to run so it makes sense to have it as a migration. I also needs to be added as the same time the OPTIONS are added to the settings.

**Comment by jonespm** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- dashboard/migrations/0020_utf8_conversion.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ITS_Teaching_And_Learning/my-learning-analytics/pullRequest?prid=5895042&bid=19303711)

**Comment by pushyamig** [comment]
I was able to run this locally applying migrations with existing data and a clean wipe of the DB Both seems to work fine. Data loaded in view and cron runs fine[](url)
[show_table_status.pdf](https://github.com/tl-its-umich-edu/my-learning-analytics/files/4960867/show_table_status.pdf)

The above pdf is create for `show table status` that is showing all the tables attributes.

Is their a way to check for individual table column status for encoding type. from the Migration file I can see individual table columns are also collated with `utf8mb4_unicode_ci`?

File: dashboard/settings.py [hunk 1]
```diff
@@ -191,6 +191,7 @@
         'PASSWORD': ENV.get('MYSQL_PASSWORD', 'student_dashboard_password'), # password for user
         'HOST': ENV.get('MYSQL_HOST', 'localhost'),
         'PORT': ENV.get('MYSQL_PORT', 3306),
+        'OPTIONS': {'charset': 'utf8mb4'},
```
Comment (pushyamig) [inline dashboard/settings.py:194]: I was using `django-mysql` package ( for Mysql DB caching support for LTI )locally the warning keeps on poping up it is not encoded with `utf8mb4_unicode_ci`. Once I set this option the warning went away

**Comment by pushyamig** [comment]
One minor suggestion I feel the `0020_utf8_conversion.py` can be numbered as `0017_utf8_conversion.py` since preceding file is 16

**Comment by pushyamig** [comment]
Over all this change seems to Work!

File: dashboard/settings.py (hunk 1 above)
Comment (jonespm) [reply dashboard/settings.py:194]: Great! Yeah that package has a check for this, but it doesn't have a fix yet for it. Good to know that this fixes that.

**Comment by jonespm** [comment]
Yeah, I was thinking about the numbering. I kept conflicting with Sam on another project and wonder if we should just skip a few numbers or a PR that has migrations should just reserve a number in advance. Like for instance I have another PR on #981 that's using 0016. 

There aren't any new text columns added as part of migration 16 but this was generated before 16. 

So this one should be 16 and the one in #981 be 17. ;)

**Comment by jonespm** [comment]
Right,

`SHOW TABLE STATUS;` will give an overview of the tables.

You can query the information_schema to get column details. This one gets all the columns where the collation_name is not null in student_dashboard

```
SELECT * FROM information_schema.columns
WHERE COLLATION_NAME IS NOT NULL
AND TABLE_SCHEMA = 'student_dashboard'
ORDER BY table_name,ordinal_position
```

Thanks for checking

**Comment by jonespm** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- dashboard/migrations/0016_utf8_conversion.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ITS_Teaching_And_Learning/my-learning-analytics/pullRequest?prid=5895042&bid=19303711)

**Comment by pushyamig** [comment]
I see one issue with this, last time when I tested it i did not notice your instructions [here](https://github.com/tl-its-umich-edu/my-learning-analytics/issues/989#issuecomment-662462855). My Apologies for that.

I started out by clean container history and deleted `.data` file. I see 2 error 
It looks to me like the site migrations and cronjob migrations must be applied first before you run this `Applying dashboard.0016_utf8_conversion`?
```
Running migrations:
student_dashboard  |   Applying contenttypes.0001_initial... OK
student_dashboard  |   Applying auth.0001_initial... OK
student_dashboard  |   Applying admin.0001_initial... OK
student_dashboard  |   Applying admin.0002_logentry_remove_auto_add... OK
student_dashboard  |   Applying admin.0003_logentry_add_action_flag_choices... OK
student_dashboard  |   Applying contenttypes.0002_remove_content_type_name... OK
student_dashboard  |   Applying auth.0002_alter_permission_name_max_length... OK
student_dashboard  |   Applying auth.0003_alter_user_email_max_length... OK
student_dashboard  |   Applying auth.0004_alter_user_username_opts... OK
student_dashboard  |   Applying auth.0005_alter_user_last_login_null... OK
student_dashboard  |   Applying auth.0006_require_contenttypes_0002... OK
student_dashboard  |   Applying auth.0007_alter_validators_add_error_messages... OK
student_dashboard  |   Applying auth.0008_alter_user_username_max_length... OK
student_dashboard  |   Applying auth.0009_alter_user_last_name_max_length... OK
student_dashboard  |   Applying auth.0010_alter_group_name_max_length... OK
student_dashboard  |   Applying auth.0011_update_proxy_permissions... OK
student_dashboard  |   Applying dashboard.0001_initial... OK
student_dashboard  |   Applying dashboard.0002_user_enrollment_type... OK
student_dashboard  |   Applying dashboard.0003_auto_20190613_1501... OK
student_dashboard  |   Applying dashboard.0004_auto_20190613_1441... OK
student_dashboard  |   Applying dashboard.0005_auto_20190613_1502... OK
student_dashboard  |   Applying dashboard.0006_auto_20190626_1217... OK
student_dashboard  |   Applying dashboard.0007_auto_20190626_1218... OK
student_dashboard  |   Applying dashboard.0008_auto_20190715_1504... OK
student_dashboard  |   Applying dashboard.0009_auto_20190731_1518... OK
student_dashboard  |   Applying dashboard.0010_auto_20190812_0924... OK
student_dashboard  |   Applying dashboard.0011_auto_20190920_1057... OK
student_dashboard  |   Applying dashboard.0012_submission_grade_posted_local_date... OK
student_dashboard  |   Applying dashboard.0013_course_show_grade_type... OK
student_dashboard  |   Applying dashboard.0014_add_auto_field_to_resources... OK
student_dashboard  |   Applying dashboard.0015_auto_20200116_1408... OK
student_dashboard  |   Applying dashboard.0016_utf8_conversion...Traceback (most recent call last):
student_dashboard  |   File "/usr/local/lib/python3.8/site-packages/django/db/backends/utils.py", line 82, in _execute
student_dashboard  |     return self.cursor.execute(sql)
student_dashboard  |   File "/usr/local/lib/python3.8/site-packages/django/db/backends/mysql/base.py", line 71, in execute
student_dashboard  |     return self.cursor.execute(query, args)
student_dashboard  |   File "/usr/local/lib/python3.8/site-packages/MySQLdb/cursors.py", line 209, in execute
student_dashboard  |     res = self._query(query)
student_dashboard  |   File "/usr/local/lib/python3.8/site-packages/MySQLdb/cursors.py", line 315, in _query
student_dashboard  |     db.query(q)
student_dashboard  |   File "/usr/local/lib/python3.8/site-packages/MySQLdb/connections.py", line 239, in query
student_dashboard  |     _mysql.connection.query(self, query)
student_dashboard  | MySQLdb._exceptions.ProgrammingError: (1146, "Table 'student_dashboard.django_cron_cronjoblog' doesn't exist")
django.db.utils.ProgrammingError: (1146, "Table 'student_dashboard.django_site' doesn't exist")
```

**Comment by jonespm** [comment]
Thanks, I didn't run this. I think I'll have to mark those scripts as dependencies. I'll fix this.

**Comment by pushyamig** [comment]
This seems to be working fine as far as I have tested it

**Pull request merged**